bugfix/realtime_safeness

v8p1197 opened the pull request:
This PR enforces real-time safeness by avoiding dynamic allocations of arrays.

**Note:** the Franka-INRIA solver is not changed for two reasons:
1. We are not the original authors of the solver, so I'd like a feedback before applying such changes
2. Internally, the methods already use fixed-size arrays, so the compiler might already optimize the code in this sense

v8p1197 committed:
[FIX] Enforce real-time safeness by avoiding dynamic allocations of vectors

v8p1197 committed:
[FIX] Avoid dynamic allocation when returning a zero-vector

v8p1197 committed:
[FIX] Fix allocation of zero-vector

my-rice commented on franka_inria_inverse_dynamics_solver/src/franka_inria_inverse_dynamics_solver.cpp:
By adding the following compile flag option to the CMakeLists.txt file, the Franka INRIA inverse dynamics solver appears to work correctly when used with real robots (and not only in simulation). Computation times seem to improve 3 times, and the runtime computation of the Coriolis torques is reduced by hundreds times compared to the default configuration.

```CMake
set(CMAKE_CXX_FLAGS "-O3")
```

However, further investigation is needed to assess the potential consequences of this optimization. It would also be valuable to experiment with other optimization levels, such as -O1, -O2, and so on. The complete list of supported optimization flags for Clang can be found [here](https://clang.llvm.org/docs/CommandGuide/clang.html).

v8p1197 committed:
[REF] Optimizes code for computation time performance

v8p1197 committed:
Merge branch 'main' into bugfix/realtime_safeness

eferre89 commented on ur10_inverse_dynamics_solver/src/ur10_inverse_dynamics_solver.cpp:
```diff
@@ -20,64 +20,60 @@
+  H_ = (double*)malloc(NUMBER_OF_JOINTS * NUMBER_OF_JOINTS * sizeof(double));
+  c_ = (double*)malloc(NUMBER_OF_JOINTS * sizeof(double));
+  g_ = (double*)malloc(NUMBER_OF_JOINTS * sizeof(double));
+  currents_ = (double*)malloc(NUMBER_OF_JOINTS * sizeof(double));
```
I think using `new` instead of `malloc` is a best practice in C++.

eferre89 commented on ur10_inverse_dynamics_solver/src/ur10_inverse_dynamics_solver.cpp:
```diff
@@ -20,64 +20,60 @@
 Eigen::MatrixXd InverseDynamicsSolverUR10::getInertiaMatrix(const Eigen::VectorXd& joint_positions) const
 {
-  double q[NUMBER_OF_JOINTS];
-  double B[NUMBER_OF_JOINTS * NUMBER_OF_JOINTS];
```
This was stack allocation, which should be associated with negligible times in C++. Have you measured the performance gain in this case?

v8p1197 replied:
| Commit                                     | Avg. Franka | Avg. KDL | Avg. UR10 | Max. Franka | Max. KDL | Max. UR10 |
| :----------------------------------------: | :---------------: | :-------: | :-------: | :---------------: | :-------: | :-------: |
| f41cd3eff335cebc78ec82f1f518601f965bbeee   | 0\.015844         | 0\.003909 | 0\.004564 | 0\.157219         | 0\.157659 | 0\.155539 |
| 24aeb34066e73bedf9bf17cdc53ecf00dfd29d98   | 0\.018625         | 0\.005732 | 0\.003934 | 0\.314219         | 0\.180907 | 0\.176222 |
| 401f3fae1751b0e80b254da46135cc9e24edb260   | 5\.394984         | 0\.005946 | 0\.012981 | 7\.789451         | 0\.197112 | 1\.804719 |
| 401f3fae1751b0e80b254da46135cc9e24edb260\* | 0\.015829         | 0\.005476 | 0\.004176 | 0\.182325         | 0\.182857 | 0\.059491 |

This table reports computation times in [ms].
Commits:
* f41cd3eff335cebc78ec82f1f518601f965bbeee is the latest commit of this branch (with `new` instead of `malloc`)
* 24aeb34066e73bedf9bf17cdc53ecf00dfd29d98 is the commit of this branch you already reviewed (with `malloc` instead of stack allocations)
* 401f3fae1751b0e80b254da46135cc9e24edb260 is the commit of the `main` branch against which this PR is open (non-RT-safe, stack allocations)
* 401f3fae1751b0e80b254da46135cc9e24edb260\* adds compilation optimization flags for a fair comparison

eferre89 replied:
It seems to me that dynamic allocation is not really giving a gain here... Is there anything I haven't been able to catch? There are two different aspects in our code: repeated allocations (one/one per method call) and stack/heap allocations. Why don't we perform ONE STACK allocation (at the time of object instantiation)?

v8p1197 replied:
With the latest commit, I have performed a single stack allocation with mutable `Eigen` arrays.

eferre89 commented on ur10_inverse_dynamics_solver/include/ur10_inverse_dynamics_solver/ur10_inverse_dynamics_solver.hpp:
```diff
@@ -81,14 +81,23 @@ class InverseDynamicsSolverUR10 : public inverse_dynamics_solver::InverseDynamic
+  double* H_;
+  double* c_;
+  double* g_;
+  double* currents_;
```
Where are these resources de-allocated?

v8p1197 committed:
[FIX] Deallocate variables

- [REF] Use new instead of malloc
- [ADD] Initialize pointers
- [ADD] Measure performance in terms of computation time

v8p1197 committed:
Merge branch 'main' into bugfix/realtime_safeness

v8p1197 committed:
[FIX] Avoid heap allocations with mutable variables

eferre89 committed:
[DOC] Fix inline comment

eferre89 committed:
[MAK] Add trailing underscore to private member attributes

eferre89 commented on ur10_inverse_dynamics_solver/include/ur10_inverse_dynamics_solver/ur10_inverse_dynamics_solver.hpp:
```diff
@@ -81,14 +82,14 @@ class InverseDynamicsSolverUR10 : public inverse_dynamics_solver::InverseDynamic
+  mutable Matrix6d H_;
+  mutable Vector6d c_;
+  mutable Vector6d g_;
+  mutable Vector6d currents_;
```
I think this implementation can now be ported to the kdl implementation, too.

v8p1197 committed:
[REF] Use mutable instead of pointers
